File: app/src/main/java/com/example/ama/questapp/domain/engine/QuestEngine.java
package com.example.ama.questapp.domain.engine;

import com.example.ama.questapp.data.db.model.QuestPattern;
import com.example.ama.questapp.data.db.model.UserTask;
import com.example.ama.questapp.domain.repointerface.QuestPatternsRepository;
import com.example.ama.questapp.domain.repointerface.UserTaskRepository;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Random;
import java.util.Set;

import javax.inject.Inject;
import javax.inject.Singleton;

import io.reactivex.Completable;
import io.reactivex.Observable;
import io.reactivex.Single;

@Singleton
public class QuestEngine implements QuestEngineGateway {
    private static final int PROGRESS_INCREMENT = 1;
    private static final int AUTO_ID = 0;
    private static final int START_PROGRESS = 0;

    private static final int ONCE_DEFAULT_TARGET = 1;
    private static final int COUNT_DEFAULT_TARGET = 3;

    private Random random = new Random();

    private QuestPatternsRepository questPatternsRepository;
    private UserTaskRepository userTaskRepository;

    @Inject
    public QuestEngine(QuestPatternsRepository questPatternsRepository, UserTaskRepository userTaskRepository) {
        this.questPatternsRepository = questPatternsRepository;
        this.userTaskRepository = userTaskRepository;
    }

    @Override
    public Completable onIncrementQuestProgress(UserTask userTask) {
        return userTaskRepository.updateUserTask(incrementProgress(userTask));
    }

    @Override
    public Completable addNewRandomDailyQuests(int maxCount) {
        return getQuestPatternsForUserTasks(maxCount)
                .flatMapObservable(Observable::fromIterable)
                .map(this::createUserTaskFromPattern)
                .buffer(maxCount)
                .flatMapCompletable(userTasks -> userTaskRepository.addUserTasks(userTasks));
    }

    private UserTask incrementProgress(UserTask userTask) {
        userTask.setTaskProgress(
                userTask.getTaskProgress() + PROGRESS_INCREMENT
        );
        if (userTask.getTaskTarget() == userTask.getTaskProgress()) {
            userTask.setCompleted(true);
        }
        return userTask;
    }

    private Single<List<QuestPattern>> getQuestPatternsForUserTasks(int maxCount) {
        return questPatternsRepository.getUnusedDailyQuestPatterns()
                .map(questPatterns -> selectPatterns(questPatterns, maxCount))
                .flatMap(questPatterns ->
                        questPatterns.size() == maxCount ?
                                Single.just(questPatterns) : addRepeatableQuestPatterns(questPatterns, maxCount));
    }

    private Single<List<QuestPattern>> addRepeatableQuestPatterns(List<QuestPattern> selectedPatterns, int maxCount) {
        return questPatternsRepository.getRepeatableDailyQuestPatterns()
                .map(questPatterns -> selectPatterns(questPatterns, maxCount - selectedPatterns.size()))
                .map(questPatterns -> {
                    selectedPatterns.addAll(questPatterns);
                    return selectedPatterns;
                });
    }

    private List<QuestPattern> selectPatterns(List<QuestPattern> patternList, int maxCount) {
        Set<QuestPattern> selectedSet = new HashSet<>();
        if (patternList.size() <= maxCount) {
            selectedSet.addAll(patternList);
        } else {
            while (selectedSet.size() < maxCount) {
                selectedSet.add(
                        patternList.get(random.nextInt(patternList.size() - 1))
                );
            }
        }
        return new ArrayList<>(selectedSet);
    }

    private UserTask createUserTaskFromPattern(QuestPattern pattern) {
        return new UserTask(AUTO_ID,
                pattern.getQuestId(),
                false,
                calcTargetCount(pattern),
                START_PROGRESS);
    }

    private int calcTargetCount(QuestPattern questPattern) {
        switch (questPattern.getQuestType()) {
            case ONCE:
                return ONCE_DEFAULT_TARGET;
            case COUNT:
                return COUNT_DEFAULT_TARGET;
            default:
                throw new IllegalArgumentException("QuestType not supported: "
                        + questPattern.getQuestType().getType());
        }
    }
}

File: app/src/main/java/com/example/ama/questapp/domain/engine/QuestEngine.kt
package com.example.ama.questapp.domain.engine

import com.example.ama.questapp.data.db.model.QuestPattern
import com.example.ama.questapp.data.db.model.QuestType
import com.example.ama.questapp.data.db.model.UserTask
import com.example.ama.questapp.domain.repointerface.QuestPatternsRepository
import com.example.ama.questapp.domain.repointerface.UserTaskRepository
import io.reactivex.Completable
import io.reactivex.Observable
import io.reactivex.Single
import java.util.*
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class QuestEngine @Inject constructor(
        private val questPatternsRepository: QuestPatternsRepository,
        private val userTaskRepository: UserTaskRepository
) : QuestEngineGateway {

    companion object {
        private const val PROGRESS_INCREMENT = 1
        private const val AUTO_ID = 0L
        private const val START_PROGRESS = 0

        private const val ONCE_DEFAULT_TARGET = 1
        private const val COUNT_DEFAULT_TARGET = 3
    }

    private val random = Random()

    override fun onIncrementQuestProgress(userTask: UserTask): Completable {
        return userTaskRepository.updateUserTask(incrementProgress(userTask))
    }

    override fun addNewRandomDailyQuests(maxCount: Int): Completable {
        return getQuestPatternsForUserTasks(maxCount)
                .flatMapObservable { Observable.fromIterable(it) }
                .map { this.createUserTaskFromPattern(it) }
                .buffer(maxCount)
                .flatMapCompletable { userTasks -> userTaskRepository.addUserTasks(userTasks) }
    }

    private fun incrementProgress(userTask: UserTask): UserTask {
        val updatedUserTask = userTask.copy(taskProgress = (userTask.taskProgress
                ?: 0) + PROGRESS_INCREMENT)
        return if (userTask.taskTarget === userTask.taskProgress) {
            updatedUserTask.copy(isCompleted = true)
        } else {
            updatedUserTask
        }
    }

    private fun getQuestPatternsForUserTasks(maxCount: Int): Single<List<QuestPattern>> {
        return questPatternsRepository.getUnusedDailyQuestPatterns()
                .map { questPatterns -> selectPatterns(questPatterns, maxCount) }
                .flatMap { questPatterns ->
                    if (questPatterns.size == maxCount)
                        Single.just(questPatterns)
                    else
                        addRepeatableQuestPatterns(questPatterns.toMutableList(), maxCount)
                }
    }

    private fun addRepeatableQuestPatterns(selectedPatterns: MutableList<QuestPattern>, maxCount: Int): Single<List<QuestPattern>> {
        return questPatternsRepository.getRepeatableDailyQuestPatterns()
                .map { questPatterns -> selectPatterns(questPatterns, maxCount - selectedPatterns.size) }
                .map { questPatterns ->
                    selectedPatterns.addAll(questPatterns)
                    selectedPatterns
                }
    }

    private fun selectPatterns(patternList: List<QuestPattern>, maxCount: Int): List<QuestPattern> {
        val selectedSet = HashSet<QuestPattern>()
        if (patternList.size <= maxCount) {
            selectedSet.addAll(patternList)
        } else {
            while (selectedSet.size < maxCount) {
                selectedSet.add(
                        patternList[random.nextInt(patternList.size - 1)]
                )
            }
        }
        return ArrayList(selectedSet)
    }

    private fun createUserTaskFromPattern(pattern: QuestPattern): UserTask {
        return UserTask(AUTO_ID,
                pattern.questId,
                false,
                calcTargetCount(pattern).toLong(),
                START_PROGRESS.toLong())
    }

    private fun calcTargetCount(questPattern: QuestPattern) =
            when (questPattern.questType) {
                QuestType.ONCE -> ONCE_DEFAULT_TARGET
                QuestType.COUNT -> COUNT_DEFAULT_TARGET
                else -> throw IllegalArgumentException("QuestType not supported: " + questPattern.questType)
            }
}
